File: retrocpu_test_framework/src/repo.ts
import path from "node:path";
import { fileURLToPath } from "node:url";

const here = path.dirname(fileURLToPath(import.meta.url));

/** RetroCpuEmu リポジトリルート */
export const REPO_ROOT = path.resolve(here, "../..");

/** retrocpu_test_framework ルート */
export const FRAMEWORK_ROOT = path.resolve(here, "..");

/** テスト成果物（HEX / CDB）の既定出力先 */
export const FRAMEWORK_BUILD = path.join(FRAMEWORK_ROOT, "build");

/** retrocpu_asm の dist/main */
export const ASM_DIST = path.join(REPO_ROOT, "retrocpu_asm/dist/main");

/** retrocpu_boot_monitor/mn1613/src */
export const MONITOR_SRC = path.join(
  REPO_ROOT,
  "retrocpu_boot_monitor/mn1613/src",
);

/** retrocpu_boot_monitor/build/hex（mn1613_mon.ihx / .cdb） */
export const MONITOR_HEX = path.join(
  REPO_ROOT,
  "retrocpu_boot_monitor/build/hex",
);

/** retrocpu_boot_monitor/mn1613/test（asm と同列のテスト） */
export const MONITOR_TEST = path.join(
  REPO_ROOT,
  "retrocpu_boot_monitor/mn1613/test",
);

/**
 * リポジトリ相対パスを絶対パスにする。
 * @param rel RetroCpuEmu からの相対パス
 * @returns 絶対パス
 */
export function repoPath(...rel: string[]): string {
  return path.join(REPO_ROOT, ...rel);
}

File: retrocpu_test_framework_rs/src/repo.rs
use std::path::PathBuf;

pub fn framework_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
}

pub fn repo_root() -> PathBuf {
    framework_root()
        .parent()
        .map(|p| p.to_path_buf())
        .unwrap_or_else(framework_root)
}

pub fn framework_build() -> PathBuf {
    framework_root().join("build")
}

pub fn asm_dist() -> PathBuf {
    repo_root().join("retrocpu_asm").join("dist").join("main")
}

pub fn monitor_src() -> PathBuf {
    repo_root()
        .join("retrocpu_boot_monitor")
        .join("src")
        .join("mn1613")
}

pub fn monitor_hex() -> PathBuf {
    repo_root()
        .join("retrocpu_boot_monitor")
        .join("build")
        .join("hex")
        .join("mn1613")
}

pub fn monitor_test() -> PathBuf {
    repo_root()
        .join("retrocpu_boot_monitor")
        .join("test")
        .join("mn1613")
}

pub fn repo_path(rel: &[&str]) -> PathBuf {
    let mut p = repo_root();
    for s in rel {
        p.push(s);
    }
    p
}
